spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkYearsFunction.java

```diff
@@ -0,0 +1,116 @@
+
+  @Before
+  public void useCatalog() {
+    sql("USE %s", catalogName);
+  }
+
+  @Test
+  public void testDates() {
+    Assert.assertEquals(
+        "Expected to produce 2017 - 1970 = 47",
+        47,
+        scalarSql("SELECT system.years(date('2017-12-01'))"));
+    Assert.assertEquals(
+        "Expected to produce 1970 - 1970 = 0",
+        0,
+        scalarSql("SELECT system.years(date('1970-01-01'))"));
+    Assert.assertEquals(
+        "Expected to produce 19690 - 1970 = -1",
+        -1,
+        scalarSql("SELECT system.years(date('1969-12-31'))"));
+    Assert.assertNull(scalarSql("SELECT system.years(CAST(null AS DATE))"));
+  }
+
+  @Test
+  public void testTimestamps() {
+    Assert.assertEquals(
+        "Expected to produce 2017 - 1970 = 47",
+        47,
+        scalarSql("SELECT system.years(TIMESTAMP '2017-12-01 10:12:55.038194 UTC+00:00')"));
+    Assert.assertEquals(
+        "Expected to produce 1970 - 1970 = 0",
+        0,
+        scalarSql("SELECT system.years(TIMESTAMP '1970-01-01 00:00:01.000001 UTC+00:00')"));
+    Assert.assertEquals(
+        "Expected to produce 19690 - 1970 = -1",
+        -1,
+        scalarSql("SELECT system.years(TIMESTAMP '1969-12-31 23:59:58.999999 UTC+00:00')"));
+    Assert.assertNull(scalarSql("SELECT system.years(CAST(null AS TIMESTAMP))"));
+  }
+
+  @Test
+  public void testWrongNumberOfArguments() {
+    AssertHelpers.assertThrows(
+        "Function resolution should not work with zero arguments",
+        AnalysisException.class,
+        "Function 'years' cannot process input: (): Wrong number of inputs",
+        () -> scalarSql("SELECT system.years()"));
+
+    AssertHelpers.assertThrows(
+        "Function resolution should not work with more than one argument",
+        AnalysisException.class,
+        "Function 'years' cannot process input: (date, date): Wrong number of inputs",
+        () -> scalarSql("SELECT system.years(date('1969-12-31'), date('1969-12-31'))"));
+  }
+
+  @Test
+  public void testInvalidInputTypes() {
+    AssertHelpers.assertThrows(
+        "Int type should not be coercible to date/timestamp",
+        AnalysisException.class,
+        "Function 'years' cannot process input: (int): Expected value to be date or timestamp",
+        () -> scalarSql("SELECT system.years(1)"));
+
+    AssertHelpers.assertThrows(
+        "Long type should not be coercible to date/timestamp",
+        AnalysisException.class,
+        "Function 'years' cannot process input: (bigint): Expected value to be date or timestamp",
+        () -> scalarSql("SELECT system.years(1L)"));
+  }
+
+  @Test
+  public void testThatMagicFunctionsAreInvoked() {
+    String dateValue = "date('2017-12-01')";
+    Assertions.assertThat(scalarSql("EXPLAIN EXTENDED SELECT system.years(%s)", dateValue))
+        .asString()
+        .isNotNull()
+        .contains(
+            "staticinvoke(class org.apache.iceberg.spark.functions.YearsFunction$DateToYearsFunction");
```

Comment:
nit: should we get the package and function name programmatically? Probably not a big deal as I can't imagine this changing

Reply:
Sounds more reliable, I'll try.

Reply:
Changed.